### PR TITLE
fix cuda compile error

### LINKED ISSUE #888: error: calling a __host__ function("__floorf") from a __device__ function("deformable_im2col_bilinear<float> ") is not allowed
Hi, I tried to compile mmcv-full (1.2.4) from source and during that an error occurs at step 9/40 (below). I followed the instructions on https://mmcv.readthedocs.io/en/latest/index.html and https://mmcv.readthedocs.io/en/latest/build.html, to build with full ops and cuda.

I used these packages to setup my conda env:
- pytorch 1.8 (installed via conda install pytorch torchvision torchaudio cudatoolkit=10.2 -c pytorch)
- python 3.8
- cudatoolkit 10.2

I use a GTX 1070 and I checked the Compute Capability which is 6.1.

It seems that there is some error during the compilation of the cu file using nvcc. I reinstalled Cuda 10.2 using the offical nvidia installer and tried different python and pytorch but I cannot figure out what is missing or what is wrong. Maybe someone can help me.

```
[9/40] C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc --generate-dependencies-with-compile --dependency-output R:\src\mmcv\build\temp.win-amd64-3.8\Release\mmcv/ops/csrc/pytorch\deform_conv_cuda.obj.d --use-local-env -Xcompiler /MD -Xcompiler /wd4819 -Xcompiler /wd4251 -Xcompiler /wd4244 -Xcompiler /wd4267 -Xcompiler /wd4275 -Xcompiler /wd4018 -Xcompiler /wd4190 -Xcompiler /EHsc -Xcudafe --diag_suppress=base_class_has_different_dll_interface -Xcudafe --diag_suppress=field_without_dll_interface -Xcudafe --diag_suppress=dll_interface_conflict_none_assumed -Xcudafe --diag_suppress=dll_interface_conflict_dllexport_assumed -DMMCV_WITH_CUDA -IR:\src\mmcv\mmcv\ops\csrc -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\torch\csrc\api\include -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\TH -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -IC:\Users\Voss\Anaconda3\include -IC:\Users\Voss\Anaconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" -c R:\src\mmcv\mmcv\ops\csrc\pytorch\deform_conv_cuda.cu -o R:\src\mmcv\build\temp.win-amd64-3.8\Release\mmcv/ops/csrc/pytorch\deform_conv_cuda.obj -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_61,code=sm_61
FAILED: R:/src/mmcv/build/temp.win-amd64-3.8/Release/mmcv/ops/csrc/pytorch/deform_conv_cuda.obj
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc --generate-dependencies-with-compile --dependency-output R:\src\mmcv\build\temp.win-amd64-3.8\Release\mmcv/ops/csrc/pytorch\deform_conv_cuda.obj.d --use-local-env -Xcompiler /MD -Xcompiler /wd4819 -Xcompiler /wd4251 -Xcompiler /wd4244 -Xcompiler /wd4267 -Xcompiler /wd4275 -Xcompiler /wd4018 -Xcompiler /wd4190 -Xcompiler /EHsc -Xcudafe --diag_suppress=base_class_has_different_dll_interface -Xcudafe --diag_suppress=field_without_dll_interface -Xcudafe --diag_suppress=dll_interface_conflict_none_assumed -Xcudafe --diag_suppress=dll_interface_conflict_dllexport_assumed -DMMCV_WITH_CUDA -IR:\src\mmcv\mmcv\ops\csrc -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\torch\csrc\api\include -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\TH -IC:\Users\Voss\Anaconda3\lib\site-packages\torch\include\THC "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -IC:\Users\Voss\Anaconda3\include -IC:\Users\Voss\Anaconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt" -c R:\src\mmcv\mmcv\ops\csrc\pytorch\deform_conv_cuda.cu -o R:\src\mmcv\build\temp.win-amd64-3.8\Release\mmcv/ops/csrc/pytorch\deform_conv_cuda.obj -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_ext -D_GLIBCXX_USE_CXX11_ABI=0 -gencode=arch=compute_61,code=sm_61
C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/macros/Macros.h(189): warning C4067: unexpected tokens following preprocessor directive - expected a newline
C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/macros/Macros.h(189): warning C4067: unexpected tokens following preprocessor directive - expected a newline
C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/ThreadLocalDebugInfo.h(12): warning: modifier is ignored on an enum specifier

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\ATen/record_function.h(19): warning: modifier is ignored on an enum specifier

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(83): error: calling a __host__ function("__floorf") from a __device__ function("deformable_im2col_bilinear<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(83): error: identifier "__floorf" is undefined in device code

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(84): error: calling a __host__ function("__floorf") from a __device__ function("deformable_im2col_bilinear<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(84): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("deformable_im2col_bilinear< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("deformable_im2col_bilinear< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(120): error: calling a __host__ function("__floorf") from a __device__ function("get_gradient_weight<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(120): error: identifier "__floorf" is undefined in device code

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(121): error: calling a __host__ function("__floorf") from a __device__ function("get_gradient_weight<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(121): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("get_gradient_weight< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("get_gradient_weight< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(147): error: calling a __host__ function("__floorf") from a __device__ function("get_coordinate_weight<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(147): error: identifier "__floorf" is undefined in device code

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(148): error: calling a __host__ function("__floorf") from a __device__ function("get_coordinate_weight<float> ") is not allowed

R:/src/mmcv/mmcv/ops/csrc\deform_conv_cuda_kernel.cuh(148): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("get_coordinate_weight< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: calling a __host__ function("__floorf") from a __device__ function("get_coordinate_weight< ::c10::Half> ") is not allowed

C:/Users/Voss/Anaconda3/lib/site-packages/torch/include\c10/util/Half-inl.h(31): error: identifier "__floorf" is undefined in device code

24 errors detected in the compilation of "C:/Users/Voss/AppData/Local/Temp/tmpxft_00003964_00000000-7_deform_conv_cuda.cpp1.ii".
deform_conv_cuda.cu

```

### PR DESCRIPTION
**Pull request by SFMDI**
to eliminate cuda compilation errors in the latest version of Visual Studio 16.9

Code modified to address the following issues
Fixes #888

### CODE REVIEW

**Comment by wangruohui** [comment]
Basically, this MR modify all `floor` to `floorf` to force to call the implementation for single precision, thus can avoid some problem on Windows. Based on comments from other users, this modification works and seems pass all tests. 
But I am not sure whether this change affect behavior when meeting double or half.

**Comment by ZwwWayne** [comment]
Hi @SFMDI ,
Thanks for your kind contribution. Could you sing the CLA and fix the lint?

**Comment by ZwwWayne** [comment]
> Basically, this MR modify all `floor` and `ceil` to `floorf` and `ceilf` to force to call the implementation for single precision, thus can avoid some problem on Windows. Based on comments from other users, this modification works and seems pass all tests.
> But I am not sure whether this change affect behavior when meeting double or half.

Usually DCN with FP16 training is not tried. We may arrange some benchmarks on that if necessary.

**Comment by RunningLeon** [comment]
@grimoire Could you take a look at this PR and the issue?

**Comment by grimoire** [comment]
If you are worrying about half and double precision, you may cast the tensor dtype in `deform_conv_cuda.cu` `DeformConvForwardCUDAKernelLauncher` before you feed the tensor into the kernel.

**Review by wangruohui — APPROVED** [review]

**Comment by wangruohui** [comment]
@SFMDI 
Hello, this MR is currently blocked because you haven't signed CLA . Would you please make it signed to unblock this MR? Thank you!

**Comment by SFMDI** [comment]
@wangruohui 
I just signed CLA

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,21 +1,21 @@
 exclude: ^tests/data/
 repos:
-  - repo: https://gitlab.com/pycqa/flake8.git
```
Comment (ZwwWayne) [inline .pre-commit-config.yaml:3]: This file should not be modified in this PR.

**Review by ZwwWayne — CHANGES_REQUESTED** [review]
pre-commit-config.yaml should be recovered. The lint issue should be fixed. For cpp code we use google style and it can be reformated in VSCode.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (wangruohui) [reply .pre-commit-config.yaml]: A mistake when I temporally debug on windows. just fixed.

**Review by ZwwWayne — APPROVED** [review]

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@SFMDI 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**